extensions/interactions/AlgebraicExpressionInput/protractor.js

```diff
@@ -0,0 +1,53 @@
+// Copyright 2020 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview End-to-end testing utilities for Algebraic Expression Input
+ * interaction
+ */
+
+var objects = require(process.cwd() + '/extensions/objects/protractor.js');
+
+var customizeInteraction = function() {
+  // There are no customizations.
+};
+
+var expectInteractionDetailsToMatch = async function(elem) {
+  expect(
+    await elem.element(by.tagName(
+      'oppia-interactive-algebraic-expression-input')).isPresent()
+  ).toBe(true);
+};
+
+var submitAnswer = async function(elem, answer) {
+  await objects.AlgebraicExpressionEditor(elem.element(by.tagName(
+    'oppia-interactive-algebraic-expression-input'))).setValue(answer);
+  await element(by.css('.protractor-test-submit-answer-button')).click();
+};
+
+var answerObjectType = 'AlgebraicExpression';
+
+var testSuite = [{
+  interactionArguments: [],
+  ruleArguments: ['MatchesExactlyWith', '((a+b))^(2)'],
```

Comment:
Does this set of tests cover the interaction's capabilities thoroughly? At first glace it seems a little too simple to be comprehensive

Reply:
Yes, this particular test is not enough to check _all_ of the possible inputs that the interaction can accept. But I was under the impression that, in the protractor tests, we are only supposed to test if the interaction behaves as expected when the end-user is using it. Its capabilities of handling various inputs are tested in the unit tests, which has been done thoroughly as you can seen [here](https://github.com/oppia/oppia/blob/64e08a8becc20485cb3fbd2b0f4de7f27fbd14d9/extensions/interactions/AlgebraicExpressionInput/directives/algebraic-expression-input-rules.service.spec.ts). Please let me know if I'm misunderstanding and if I should add all those tests here as well.

Reply:
You are correct that e2e tests don't need to be quite as thorough as the unit tests, but they still need to be fairly comprehensive because the kinds of bugs we find with e2e tests can be just as subtle as those found by unit tests. For example, we recently discovered a bug with the exploration editor that only occurred when a drag-and-drop interaction was present. This wasn't caught by the e2e tests (or any of our tests for that matter) because the exploration editor tests aren't comprehensive enough and only test a few of the most common interactions, not drag-and-drop. To catch bugs like these, ideally we'd have e2e tests that cover all possible ways the end-user could use the site.

However, it's also true that it's impractical to actually test every possible usage. To take an extreme, we'd never consider testing all possible topic names in case some special sequence of characters caused it to crash (think of that [iPhone-crashing string](https://www.cnet.com/news/new-ios-text-bomb-bug-can-crash-your-iphone-but-fix-is-said-to-be-close/) from a few months ago). I think the `ItemSelectionInput` test strikes a reasonable middle ground:

https://github.com/oppia/oppia/blob/ddc50f2ff4e5cb9fc4d337f7c3289c50532ee382/extensions/interactions/ItemSelectionInput/protractor.js#L118

Does that sound reasonable?

Reply:
Thanks for the clear explanation! I've added more tests.

Reply:
Thanks!